packages/kbn-test/src/es/test_es_cluster.ts

```diff
@@ -88,31 +151,51 @@ export function createTestEsCluster(options: TestClusterFactoryOptions) {
     }
 
     async start() {
-      let installPath;
+      let installPath: string;
 
+      // We only install once using the first node. If the cluster has
+      // multiple nodes, they'll all share the same ESinstallation.
+      const firstNode = this.nodes[0];
       if (esFrom === 'source') {
-        installPath = (await cluster.installSource(config)).installPath;
+        installPath = (await firstNode.installSource(config)).installPath;
       } else if (esFrom === 'snapshot') {
-        installPath = (await cluster.installSnapshot(config)).installPath;
+        installPath = (await firstNode.installSnapshot(config)).installPath;
       } else if (Path.isAbsolute(esFrom)) {
         installPath = esFrom;
       } else {
         throw new Error(`unknown option esFrom "${esFrom}"`);
       }
 
-      if (dataArchive) {
-        await cluster.extractDataDirectory(installPath, dataArchive);
-      }
+      for (let i = 0; i < this.nodes.length; i++) {
+        const node = nodes[i];
+        const nodePort = this.ports[i];
+        const overriddenArgs = [`node.name=${node.name}`, `http.port=${nodePort}`];
 
-      await cluster.start(installPath, {
-        password: config.password,
-        esArgs,
-        esJavaOpts,
-      });
+        const archive = node.dataArchive || dataArchive;
+        if (archive) {
+          const nodeDataDirectory = node.dataArchive ? `data-${node.name}` : 'data';
+          await this.nodes[i].extractDataDirectory(installPath, archive, nodeDataDirectory);
+          overriddenArgs.push(`path.data=${Path.resolve(installPath, nodeDataDirectory)}`);
+        }
+
+        log.info(`[es] starting node ${node.name} on port ${nodePort}`);
+        await this.nodes[i].start(installPath, {
+          password: config.password,
+          esArgs: assignArgs(esArgs, overriddenArgs),
+          esJavaOpts,
+          // If we have multiple nodes, we shouldn't try setting up the native realm
+          // right away, or ES will complain as the cluster isn't ready. So we only
+          // set it up after the last node is started.
+          skipNativeRealmSetup: this.nodes.length > 1 && i < this.nodes.length - 1,
```

Comment:
It looks like a `NativeRealm` setup should be decoupled from `start` . How hard it would be to move it outside of `start` lifecycle?

Reply:
Yeah I agree that does feel like the right approach. I was intentionally avoiding moving too much around in `Cluster`, but I'll take a closer look.

I think `createTestEsCluster` is the only place using `Cluster` directly, so probably wouldn't be too hard to split out into a separate step, e.g.
```js
await nodeA.start();
await nodeB.start();
await setupNativeRealm({ port, password });
```

It's cleaner having them decoupled, but also one more step to take that most people will want to do by default. However this is probably okay as I don't envision anyone needing to use `Cluster` directly, just `createTestEsCluster`.

Reply:
Really what would be a nicer long-term design is breaking `Cluster` apart into a `Node` class which is similar to the current implementation, and then a higher-level `Cluster` abstraction that would coordinate starting/stopping all nodes and then performing setup tasks like native realm.

I think that's a little far outside the scope of this PR, but I think would be worth considering down the road.

Reply:
@lukeelmers ok, let's create an issue describing the problem?

Reply:
I started decoupling `NativeRealm` from `start`. It works fine with `cluster.start`, as that promise resolves once ES has started.

However, I had not noticed there's `cluster.run`, which does not resolve until ES exits. That makes separating native realm setup into a different step problematic, as you won't know when ES is "ready" for the native realm setup.

I went ahead and reverted those changes, leaving `skipNativeRealmSetup` untouched. I can document this in the follow-up issue as well, since it's something that'd be nice to resolve eventually -- I still don't like the `skipNativeRealmSetup` hack and would rather not keep it around forever.